clippy_lints/src/suspicious_arguments.rs

```diff
@@ -0,0 +1,140 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::path_res;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for calls to a function where the parameters look swapped
+    ///
+    /// ### Why is this bad?
+    ///
+    /// This likely indicates an error, where the arguments were reversed.
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn resize(width: usize, height: usize) {}
+    ///
+    /// let height = 100;
+    /// let width = 200;
+    /// resize(height, width);
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn resize(width: usize, height: usize) {}
+    ///
+    /// let height = 100;
+    /// let width = 200;
+    /// resize(width, height);
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub SUSPICIOUS_ARGUMENTS,
+    suspicious,
+    "function call with probably swapped arguments"
+}
+declare_lint_pass!(SuspiciousArguments => [SUSPICIOUS_ARGUMENTS]);
+
+fn arguments_are_sus(cx: &LateContext<'_>, definition: &[(String, Span)], call: &[Option<(String, Span)>]) {
+    let idxs: FxHashMap<&String, usize> = definition
+        .iter()
+        .enumerate()
+        .map(|(idx, (item, _))| (item, idx))
+        .collect();
+
+    for (call_idx, arg_and_span) in call.iter().enumerate() {
+        if let Some((arg, call_span)) = arg_and_span {
+            if let Some(&def_idx) = idxs.get(arg) {
+                if call_idx != def_idx {
+                    if let Some((reverse_call, reverse_call_span)) = &call[def_idx] {
+                        let def_for_call = &definition[call_idx];
+                        if reverse_call == &def_for_call.0 {
```

Comment:
There is quite some rightwards drift here. We can use either the `if_chain` macro or the recently reverted-on-stable-but-available-on-nightly let-chains feature (basically `if let Some((arg, call_span)) = arg_and_span && let Some(&def_idx) = idxs.get(arg) && call_idx != def_idx && ..` etc.).